django/core/cache/backends/memcached.py

```diff
@@ -213,3 +213,55 @@ def _modify_servers(servers):
                 yield server[5:]
             else:
                 yield server
+
+
+class PyMemcacheCache(BaseMemcachedCache):
+    """An implementation of a cache binding using pymemcache."""
+
+    def __init__(self, server, params):
+        import pymemcache.client.hash
+        import pymemcache.serde
+        super().__init__(server, params, library=pymemcache, value_not_found_exception=KeyError)
+        self._class = type('PyMemcacheClient', (PyMemcacheCompatMixin, self._lib.client.hash.HashClient), {})
+        self._options = {
+            'allow_unicode_keys': True,
+            'default_noreply': False,
+            'serde': pymemcache.serde.pickle_serde,
+            **self._options,
+        }
+
+    def touch(self, key, timeout=DEFAULT_TIMEOUT, version=None):
+        key = self.make_key(key, version=version)
+        return self._cache.touch(key, self.get_backend_timeout(timeout))
+
+
+class PyMemcacheCompatMixin:
+    """A mixin to work around API incompatibilities in pymemcache clients."""
+
+    def __init__(self, servers, *args, **kwargs):
+        servers = list(self._modify_servers(servers))
+        super().__init__(servers, *args, **kwargs)
+
+    @staticmethod
+    def _modify_servers(servers):
+        for server in servers:
+            if server.startswith('unix:'):
+                yield server[5:]
+            elif server.startswith('/'):
+                yield server
+            elif ':' not in server or server.endswith(']'):
```

Comment:
Although `pymemcache` doesn't currently support IPv6, it makes sense to check for IPv6 addresses here to avoid splitting incorrectly. Support may be added in the future, see pinterest/pymemcache#257.

Reply:
Can we use `is_valid_ipv6_address()`?